### DISCUSSION TITLE
self-hosted supabase realtime not working using offical docker guide ,error showing while connecting to channels from dashboard Failed to connect to the channel: This may be due to restrictive RLS policies. Check your role and try again.

### DISCUSSION THREAD

**Discussion by boyhax** [supabase/supabase#26749] [Questions]
please explain me why no one talks about this issue because i tried alot to find solution for last 5 days .
im trying to self host supabase in hosting vps 1 core 4gb ram .im using offical docker guides and also tried so many things like changing os from debian to almalinux to ubuntu and nothing changes .i also tested different realtime versions but nothing changed.
the realtime log shows connection failed but still dont explain more .i tried different ports nothing helped .also same issue effect js client side where realtime subscribe status is always "channel closed" or something like that but at the end its same issue .   

realtime log in dashboard not working also.

realtime log after channel connecting attempt from dashboard .
18:13:39.166 [warning] Replica region not found, defaulting to Realtime.Repo
18:13:39.167 [debug] QUERY OK source="tenants" db=1.4ms idle=1241.5ms
SELECT t0."id", t0."name", t0."external_id", t0."jwt_secret", t0."jwt_jwks", t0."postgres_cdc_default", t0."max_concurrent_users", t0."max_events_per_second", t0."max_bytes_per_second", t0."max_channels_per_client", t0."max_joins_per_second", t0."suspend", t0."enable_authorization", t0."inserted_at", t0."updated_at" FROM "tenants" AS t0 WHERE (t0."external_id" = $1) ["realtime-dev"]
18:13:39.168 [debug] QUERY OK source="extensions" db=0.8ms idle=1242.9ms
SELECT e0."id", e0."type", e0."settings", e0."tenant_external_id", e0."inserted_at", e0."updated_at", e0."tenant_external_id" FROM "extensions" AS e0 WHERE (e0."tenant_external_id" = $1) ORDER BY e0."tenant_external_id" ["realtime-dev"]
18:13:39.170 request_id=F9J_VuTiZG6Z7F4AABKx project=realtime-dev external_id=realtime-dev [info] Sent 200 in 4ms
18:13:44.254 request_id=F9J_WBQY-tA_be8AABLB [info] HEAD /api/tenants/realtime-dev/health
18:13:44.254 request_id=F9J_WBQY-tA_be8AABLB [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:13:44.258 request_id=F9J_WBQY-tA_be8AABLB project=realtime-dev external_id=realtime-dev [info] Sent 200 in 5ms
18:13:49.314 request_id=F9J_WUHPmdh8jAIAABLR [info] HEAD /api/tenants/realtime-dev/health
18:13:49.314 request_id=F9J_WUHPmdh8jAIAABLR [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:13:49.319 request_id=F9J_WUHPmdh8jAIAABLR project=realtime-dev external_id=realtime-dev [info] Sent 200 in 4ms
18:13:54.368 request_id=F9J_Wm8JCiviBBoAABLh [info] HEAD /api/tenants/realtime-dev/health
18:13:54.368 request_id=F9J_Wm8JCiviBBoAABLh [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:13:54.370 request_id=F9J_Wm8JCiviBBoAABLh project=realtime-dev external_id=realtime-dev [info] Sent 200 in 2ms
18:13:59.435 request_id=F9J_W50TGpoUUtsAABLx [info] HEAD /api/tenants/realtime-dev/health
18:13:59.435 request_id=F9J_W50TGpoUUtsAABLx [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:13:59.438 request_id=F9J_W50TGpoUUtsAABLx project=realtime-dev external_id=realtime-dev [info] Sent 200 in 2ms
18:14:04.501 request_id=F9J_XMsB0D-vq8oAABMB [info] HEAD /api/tenants/realtime-dev/health
18:14:04.501 request_id=F9J_XMsB0D-vq8oAABMB [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:04.504 request_id=F9J_XMsB0D-vq8oAABMB project=realtime-dev external_id=realtime-dev [info] Sent 200 in 3ms
18:14:09.545 request_id=F9J_XfewcqjVttwAABMR [info] HEAD /api/tenants/realtime-dev/health
18:14:09.546 request_id=F9J_XfewcqjVttwAABMR [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:09.546 [warning] Replica region not found, defaulting to Realtime.Repo
18:14:09.547 [debug] QUERY OK source="tenants" db=1.2ms idle=1621.7ms
SELECT t0."id", t0."name", t0."external_id", t0."jwt_secret", t0."jwt_jwks", t0."postgres_cdc_default", t0."max_concurrent_users", t0."max_events_per_second", t0."max_bytes_per_second", t0."max_channels_per_client", t0."max_joins_per_second", t0."suspend", t0."enable_authorization", t0."inserted_at", t0."updated_at" FROM "tenants" AS t0 WHERE (t0."external_id" = $1) ["realtime-dev"]
18:14:09.548 [debug] QUERY OK source="extensions" db=0.4ms queue=0.4ms idle=1622.9ms
SELECT e0."id", e0."type", e0."settings", e0."tenant_external_id", e0."inserted_at", e0."updated_at", e0."tenant_external_id" FROM "extensions" AS e0 WHERE (e0."tenant_external_id" = $1) ORDER BY e0."tenant_external_id" ["realtime-dev"]
18:14:09.550 request_id=F9J_XfewcqjVttwAABMR project=realtime-dev external_id=realtime-dev [info] Sent 200 in 4ms
18:14:14.602 request_id=F9J_XyUVjsFvDroAABMh [info] HEAD /api/tenants/realtime-dev/health
18:14:14.602 request_id=F9J_XyUVjsFvDroAABMh [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:14.605 request_id=F9J_XyUVjsFvDroAABMh project=realtime-dev external_id=realtime-dev [info] Sent 200 in 3ms
18:14:19.649 request_id=F9J_YFHuVU1MnyYAABMx [info] HEAD /api/tenants/realtime-dev/health
18:14:19.649 request_id=F9J_YFHuVU1MnyYAABMx [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:19.652 request_id=F9J_YFHuVU1MnyYAABMx project=realtime-dev external_id=realtime-dev [info] Sent 200 in 2ms
18:14:24.708 request_id=F9J_YX95P-mePxkAABNB [info] HEAD /api/tenants/realtime-dev/health
18:14:24.709 request_id=F9J_YX95P-mePxkAABNB [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:24.711 request_id=F9J_YX95P-mePxkAABNB project=realtime-dev external_id=realtime-dev [info] Sent 200 in 3ms
18:14:29.757 request_id=F9J_YqxnOdW3cKUAABNR [info] HEAD /api/tenants/realtime-dev/health
18:14:29.757 request_id=F9J_YqxnOdW3cKUAABNR [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:29.759 request_id=F9J_YqxnOdW3cKUAABNR project=realtime-dev external_id=realtime-dev [info] Sent 200 in 2ms
18:14:34.800 request_id=F9J_Y9j9N0-x760AABNh [info] HEAD /api/tenants/realtime-dev/health
18:14:34.800 request_id=F9J_Y9j9N0-x760AABNh [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:34.803 request_id=F9J_Y9j9N0-x760AABNh project=realtime-dev external_id=realtime-dev [info] Sent 200 in 2ms
18:14:39.876 request_id=F9J_ZQeIayOmyFoAABNx [info] HEAD /api/tenants/realtime-dev/health
18:14:39.876 request_id=F9J_ZQeIayOmyFoAABNx [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:39.876 [warning] Replica region not found, defaulting to Realtime.Repo
18:14:39.882 [debug] QUERY OK source="tenants" db=5.0ms idle=1951.4ms
SELECT t0."id", t0."name", t0."external_id", t0."jwt_secret", t0."jwt_jwks", t0."postgres_cdc_default", t0."max_concurrent_users", t0."max_events_per_second", t0."max_bytes_per_second", t0."max_channels_per_client", t0."max_joins_per_second", t0."suspend", t0."enable_authorization", t0."inserted_at", t0."updated_at" FROM "tenants" AS t0 WHERE (t0."external_id" = $1) ["realtime-dev"]
18:14:39.883 [debug] QUERY OK source="extensions" db=0.3ms queue=0.4ms idle=956.5ms
SELECT e0."id", e0."type", e0."settings", e0."tenant_external_id", e0."inserted_at", e0."updated_at", e0."tenant_external_id" FROM "extensions" AS e0 WHERE (e0."tenant_external_id" = $1) ORDER BY e0."tenant_external_id" ["realtime-dev"]
18:14:39.884 request_id=F9J_ZQeIayOmyFoAABNx project=realtime-dev external_id=realtime-dev [info] Sent 200 in 8ms
18:14:44.935 request_id=F9J_ZjUQNZzJMLwAABOB [info] HEAD /api/tenants/realtime-dev/health
18:14:44.935 request_id=F9J_ZjUQNZzJMLwAABOB [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:44.937 request_id=F9J_ZjUQNZzJMLwAABOB project=realtime-dev external_id=realtime-dev [info] Sent 200 in 2ms
18:14:49.981 request_id=F9J_Z2HS_mNcur4AABOR [info] HEAD /api/tenants/realtime-dev/health
18:14:49.981 request_id=F9J_Z2HS_mNcur4AABOR [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:49.984 request_id=F9J_Z2HS_mNcur4AABOR project=realtime-dev external_id=realtime-dev [info] Sent 200 in 3ms
18:14:54.878 project=realtime-dev external_id=realtime-dev application_name=realtime_connect [info] Tenant has no connected users, database connection will be terminated
18:14:55.029 request_id=F9J_aI66BKLTG5YAABOh [info] HEAD /api/tenants/realtime-dev/health
18:14:55.029 request_id=F9J_aI66BKLTG5YAABOh [debug] Processing with RealtimeWeb.TenantController.health/2
  Parameters: %{"tenant_id" => "realtime-dev"}
  Pipelines: [:api]
18:14:55.030 request_id=F9J_aI66BKLTG5YAABOh project=realtime-dev external_id=realtime-dev [warning] Zero region nodes for iad using :"realtime@127.0.0.1"
18:14:55.031 project=realtime-dev external_id=realtime-dev [warning] Replica region not found, defaulting to Realtime.Repo
18:14:55.032 project=realtime-dev external_id=realtime-dev [debug] QUERY OK source="tenants" db=0.4ms queue=0.3ms idle=105.9ms
SELECT t0."id", t0."name", t0."external_id", t0."jwt_secret", t0."jwt_jwks", t0."postgres_cdc_default", t0."max_concurrent_users", t0."max_events_per_second", t0."max_bytes_per_second", t0."max_channels_per_client", t0."max_joins_per_second", t0."suspend", t0."enable_authorization", t0."inserted_at", t0."updated_at" FROM "tenants" AS t0 WHERE (t0."external_id" = $1) ["realtime-dev"]
18:14:55.033 project=realtime-dev external_id=realtime-dev [debug] QUERY OK source="extensions" db=0.4ms queue=0.4ms idle=106.8ms
SELECT e0."id", e0."type", e0."settings", e0."tenant_external_id", e0."inserted_at", e0."updated_at", e0."tenant_external_id" FROM "extensions" AS e0 WHERE (e0."tenant_external_id" = $1) ORDER BY e0."tenant_external_id" ["realtime-dev"]
18:14:55.080 project=realtime-dev external_id=realtime-dev application_name=realtime_connect [info] Applying migrations to db
18:14:55.116 project=realtime-dev external_id=realtime-dev application_name=realtime_connect [info] Migrations already up
18:14:55.121 request_id=F9J_aI66BKLTG5YAABOh project=realtime-dev external_id=realtime-dev [info] Sent 200 in 92ms

**Comment by sinameraji** [comment]
in your server, when you run docker compose ps, is supabase-realtime's status healthy or unhealthy?

if it's unhealthy, can you run docker logs supabase-realtime and docker inspect supabase-realtime and share the responses?

**Comment by Diego-SJ** [comment]
any solution? I'm having the same issue in 2025, in my ocker container have a healthy status, but I'm getting

```
Failed to connect to the channel test: This may be due to restrictive RLS policies. Check your role and try again.
```

when try to join channel

**Comment by Coop4Free** [comment]
This is super annoying and no one is talking about it. Please fix that!

**Comment by gdut-yy** [comment]
This is super annoying and no one is talking about it. Please fix that!

**Reply by Audiojoy72** [reply]
setting up self hosted supabase is a pain i.d.a.
i struggling since months with it trying to understand and get it to work save, but i dont trust it. the next time i do little tweaks in the config or trying to set it up on another instance i m getting new and different issues, errors and health check problems. and the docu is also not really helpful and somethings wrong

**Comment by peekknuf** [comment]
Hi! I'm going to chime in and make a couple of assumptions and maybe this leads somewhere

1. You use private channels (as you should)
2. You successfully granted RLS policies word for word as it is indicated in the [Getting Started](https://supabase.com/docs/guides/realtime/getting_started#4-set-up-authorization), if not, you must.
3. `restrictive RLS policies` error, might not be intuitive, but it's a call to first set those policies up
4. Still issues like this one (at least i had this)
   `Error: "Unauthorized: You do not have permissions to read from this Channel topic: room:lobby:messages"`

If that's the case, by default when you're self-hosting and developing locally you're using jwt with embedded anon role, the thing is, if you copy pasted the RLS in the documentation you've granted the access to the **authenticated** role, when you're **anon** one.



Once you grant the RLS policies to the role you're developing with like this:
```
CREATE POLICY "anon_users_can_receive" ON realtime.messages
    FOR SELECT TO anon USING (true);
CREATE POLICY "anon_users_can_send" ON realtime.messages
    FOR INSERT TO anon WITH CHECK (true);
```
This should work. If anything, that's a move in _some_ direction

P.S You can check your role by decoding your JWT at https://jwt.io/ - look for the 'role' in Decoded Payload
